### PR TITLE
feat(HACBS-682): only create release if releaselink set to auto-release

### PR DESCRIPTION
**Pull request by johnbieren**
This commit adds functionality to the integration_adapter to only retrieve the ReleaseLinks
for an application if the ReleaseLink does not have the release.appstudio.openshift.io/auto-release
label set to false. A missing label is considered the same as it being there and set to true.

Signed-off-by: Johnny Bieren <jbieren@redhat.com>

### CODE REVIEW

**Comment by johnbieren** [comment]
@davidmogar untested so it is marked as draft, but FYI. I am going to open a PR for the ReleaseLink webhook to disallow setting the `auto-release` label to anything but true/false. Otherwise this code breaks if we set it to `foo` or something.

**Comment by johnbieren** [comment]
https://github.com/redhat-appstudio/release-service/pull/63 forces updates to the release link value to be only true or false

File: controllers/integration/integration_adapter.go [hunk 1]
```diff
@@ -189,15 +193,24 @@ func (a *Adapter) getAllApplicationComponents(application *hasv1alpha1.Applicati
...
-// ReleaseLinks are not found, an error will be returned.
+// ReleaseLinks are not found, an error will be returned. A ReleaseLink will only be returned if it has the
+// release.appstudio.openshift.io/auto-release label set to true or if it is missing the label entirely.
 func (a *Adapter) getAllApplicationReleaseLinks(application *hasv1alpha1.Application) (*[]releasev1alpha1.ReleaseLink, error) {
 	releaseLinks := &releasev1alpha1.ReleaseLinkList{}
-	opts := []client.ListOption{
-		client.InNamespace(application.Namespace),
-		client.MatchingFields{"spec.application": application.Name},
+	labelSelector := labels.NewSelector()
+	rlReq, err := labels.NewRequirement("release.appstudio.openshift.io/auto-release", selection.DoesNotExist, []string{"false"})
```
Comment (davidmogar) [inline controllers/integration/integration_adapter.go:201]: This won’t work. When using DoesNotExist, the value passed must be empty. Maybe you can add multiple requirements to the selector, but I’m not sure about the logical expression used in those cases. I’d say that you can use NotIn with that value, to check that it’s not set to false. I’m assuming that if the label is not present, this resolve to true as well.

**Comment by johnbieren** [comment]
@dirgim I don't see any integration controller stuff in https://github.com/redhat-appstudio/infra-deployments/blob/main/hack/preview-template.env 
How do y'all normally test out code changes?

**Comment by davidmogar** [comment]
> @dirgim I don't see any integration controller stuff in https://github.com/redhat-appstudio/infra-deployments/blob/main/hack/preview-template.env
> How do y'all normally test out code changes?

I added this https://github.com/redhat-appstudio/infra-deployments/pull/425. Just rebase it on top of your fork main and you are good to go.

**Comment by johnbieren** [comment]
> > @dirgim I don't see any integration controller stuff in https://github.com/redhat-appstudio/infra-deployments/blob/main/hack/preview-template.env
> > How do y'all normally test out code changes?
> 
> I added this [redhat-appstudio/infra-deployments#425](https://github.com/redhat-appstudio/infra-deployments/pull/425). Just rebase it on top of your fork main and you are good to go.

Thank you!

File: controllers/integration/integration_adapter.go [hunk 2]
```diff
@@ -189,15 +193,24 @@ func (a *Adapter) getAllApplicationComponents(application *hasv1alpha1.Applicati
...
-// ReleaseLinks are not found, an error will be returned.
+// ReleaseLinks are not found, an error will be returned. A ReleaseLink will only be returned if it has the
+// release.appstudio.openshift.io/auto-release label set to true or if it is missing the label entirely.
 func (a *Adapter) getAllApplicationReleaseLinks(application *hasv1alpha1.Application) (*[]releasev1alpha1.ReleaseLink, error) {
 	releaseLinks := &releasev1alpha1.ReleaseLinkList{}
-	opts := []client.ListOption{
-		client.InNamespace(application.Namespace),
-		client.MatchingFields{"spec.application": application.Name},
+	labelSelector := labels.NewSelector()
+	rlReq, err := labels.NewRequirement("release.appstudio.openshift.io/auto-release", selection.NotIn, []string{"false"})
```
[the remark this answers is not in the document]
Comment (davidmogar) [reply controllers/integration/integration_adapter.go:201]: Just by checking [the code](https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L225) this should work as expected. I haven't tested it though.

**Comment by johnbieren** [comment]
I tested this with three scenarios

- releaselink has auto-release label set to true -> release is created -> worked properly
- releaselink has auto-release label set to false -> release is not created -> worked properly
- releaselink is missing the auto-release label -> release is created -> worked properly

None of this looks at the releaselink in the managed workspace, but that is m7 stuff afaik

File: controllers/integration/integration_adapter.go [hunk 3]
```diff
@@ -189,15 +193,24 @@ func (a *Adapter) getAllApplicationComponents(application *hasv1alpha1.Applicati
...
+	labelSelector = labelSelector.Add(*labelRequirement)
+
+	opts := client.ListOptions{
+		Namespace:     application.Namespace,
+		FieldSelector: fields.OneTermEqualSelector("spec.application", application.Name),
+		LabelSelector: labelSelector,
 	}
 
-	err := a.client.List(a.context, releaseLinks, opts...)
+	err = a.client.List(a.context, releaseLinks, &opts)
```
Comment (davidmogar) [inline controllers/integration/integration_adapter.go:213]: Wondering if it makes sense to move the `&` to the line where the instance is created. We do always the same for resources declarations.

**Review by davidmogar — COMMENTED** [review]
Looking good to me.

File: controllers/integration/integration_adapter.go (hunk 3 above)
Comment (johnbieren) [reply controllers/integration/integration_adapter.go]: done

File: controllers/integration/integration_adapter.go [hunk 4]
```diff
@@ -189,15 +193,24 @@ func (a *Adapter) getAllApplicationComponents(application *hasv1alpha1.Applicati
 }
 
 // getAllApplicationReleaseLinks returns the ReleaseLinks used by the application being processed. If matching
-// ReleaseLinks are not found, an error will be returned.
+// ReleaseLinks are not found, an error will be returned. A ReleaseLink will only be returned if it has the
+// release.appstudio.openshift.io/auto-release label set to true or if it is missing the label entirely.
 func (a *Adapter) getAllApplicationReleaseLinks(application *hasv1alpha1.Application) (*[]releasev1alpha1.ReleaseLink, error) {
```
[the remark this answers is not in the document]
Comment (johnbieren) [reply controllers/integration/integration_adapter.go]: done

**Comment by johnbieren** [comment]
> > I tested this with three scenarios
> > 
> > * releaselink has auto-release label set to true -> release is created -> worked properly
> > * releaselink has auto-release label set to false -> release is not created -> worked properly
> > * releaselink is missing the auto-release label -> release is created -> worked properly
> > 
> > None of this looks at the releaselink in the managed workspace, but that is m7 stuff afaik
> 
> @johnbieren Could you share how you tested your scenarios? I would like to learn. Thank you!

I deployed `infra-deployments` with a custom integration-controller specific to my work here (We recorded how to do this. If you don't have the recording I can share it with you). I then just created a bunch of `ReleaseLinks` to fit the scenarios I wanted (over a full screen of yaml so not pasting it here unless you want me to). Then I just created dummy pipeline runs in my cluster to see what would happen. The release pipeline failed for all since I was just using dummy things, but I was only interested in seeing if the integration controller created a release or not for each pipeline run

**Review by dirgim — APPROVED** [review]
LGTM

**Review by jsztuka — APPROVED** [review]
LGTM

**Review by staticf0x — APPROVED** [review]
LGTM

**Comment by johnbieren** [comment]
@dirgim got some approvals, can we merge this?

**Comment by dirgim** [comment]
@johnbieren yes, thanks for putting this together!

**Pull request merged**